src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPropertyInfo.cs

```diff
@@ -100,24 +99,116 @@ public abstract class JsonPropertyInfo
                 _shouldSerialize = value;
                 // By default we will go through faster path (not using delegate) and use IgnoreCondition
                 // If users sets it explicitly we always go through delegate
-                IgnoreCondition = null;
+                _ignoreCondition = null;
                 IsIgnored = false;
                 _shouldSerializeIsExplicitlySet = true;
             }
         }
 
-        private protected Func<object, object?, bool>? _shouldSerialize;
+        internal JsonIgnoreCondition? IgnoreCondition
+        {
+            get => _ignoreCondition;
+            set
+            {
+                Debug.Assert(!_isConfigured);
+
+                _ignoreCondition = value;
+                IsIgnored = value == JsonIgnoreCondition.Always;
+
+                if (value != null)
+                {
+                    _shouldSerialize = GetShouldSerializeForIgnoreCondition(value.Value);
+                }
```

Comment:
```csharp
else
{
    _shouldSerialize = null;
}
```

Reply:
Good call, although I think currently we might get away with just asserting `Debug.Assert(_shouldSerialize == null);` since it's an internal property set at initialization. Might be worth doing it in case we ever publish this property though.